File: 1368.minimum-cost-to-make-at-least-one-valid-path-in-a-grid.go
package leetcode

import "math"

// @leet start
type coord struct {
	i, j int
}

func minCost(grid [][]int) int {
	n := len(grid)
	m := len(grid[0])
	dirs := map[int]coord{
		1: {0, 1},  // right
		2: {0, -1}, // left
		3: {1, 0},  // down
		4: {-1, 0}, // up
	}

	// 0-1 bfs
	root := coord{0, 0}

	dists := make([][]int, n)
	for i := range dists {
		dists[i] = make([]int, m)
		for j := range dists[i] {
			dists[i][j] = math.MaxInt
		}
	}
	dists[0][0] = 0

	queue := []coord{root}

	for len(queue) > 0 {
		curr := queue[0]
		queue = queue[1:]

		for dir := 1; dir <= 4; dir++ {
			i, j := curr.i, curr.j
			di, dj := dirs[dir].i, dirs[dir].j
			ni, nj := i+di, j+dj

			if 0 <= ni && ni < n && 0 <= nj && nj < m {
				cost := 0
				if dir != grid[i][j] {
					cost = 1
				}

				if dists[i][j]+cost < dists[ni][nj] {
					dists[ni][nj] = dists[i][j] + cost
					next := coord{ni, nj}
					if cost == 0 {
						queue = append([]coord{next}, queue...) // prepend
					} else {
						queue = append(queue, next) // append
					}
				}
			}
		}
	}

	return dists[n-1][m-1]

}

// @leet end

File: 1368.minimum-cost-to-make-at-least-one-valid-path-in-a-grid.py
from collections import defaultdict
from heapq import heappop, heappush
from typing import List


# @leet start
class Solution:

    def minCost(self, grid: List[List[int]]) -> int:
        # The key point is that this is actually a graph question...
        # TC: O(V + E logV) = O(nlogn)
        # SC: O(n) for pq, dists
        # apparently using a deque in 0-1 bfs is more efficient

        n = len(grid)
        m = len(grid[0])

        dirs = {
            1: (0, 1),  # right
            2: (0, -1),  # left
            3: (1, 0),  # down
            4: (-1, 0),  # up
        }

        # construct a weighted graph with correct dir weight 0 and wrong dir weight 1
        graph = defaultdict(lambda: defaultdict(int))

        for i in range(n):
            for j in range(m):
                for dir, (di, dj) in dirs.items():
                    ni = i + di
                    nj = j + dj

                    if 0 <= ni < n and 0 <= nj < m:
                        graph[(i, j)][(ni, nj)] = 0 if dir == grid[i][j] else 1

        # dijkstra
        root = (0,0)
        pq = [root]
        dists = defaultdict(lambda: n * m)
        dists[root] = 0
        while pq:
            curr = heappop(pq)
            if curr == (n - 1, m - 1):
                return dists[(n - 1, m - 1)]

            nbs = graph[curr].items()
            for nb, weight in nbs:
                if dists[curr] + weight < dists[nb]:
                    dists[nb] = dists[curr] + weight
                    heappush(pq, nb)

        return -1


# @leet end

grid = [[1, 1, 1, 1], [2, 2, 2, 2], [1, 1, 1, 1], [2, 2, 2, 2]]
x = Solution().minCost(grid)
print(x)

